### PR TITLE
1111 add damerau levenshtein

### PR DESCRIPTION
**Pull request by RossKen**
Add Damerau-Levenshtein distance to cl and cll libraries.

Massive kudos to 
@mamonu for the scala UDF
@ADBond for his PR (not merged yet) into Duckdb

Once the duckdb function is integrated I will take this out of draft mode for review.

Note: this branch has changes from #1167 merged in as well to make use of some more general functions not yet in master

EDIT DuckDB PR merged, waiting for next release

### CODE REVIEW

**Comment by RossKen** [comment]
Damerau - Levenshtein function merged into master in DuckDB - waiting for next release then this should be good to go

**Comment by RossKen** [comment]
This is good for review. Couple of points:
* Splink demos are failing due to a previous merged PR (I will update when we do our next release)
* There are some changes to the default comparator and thresholds chosen for the ctl functions based on D-L being available

File: pyproject.toml [hunk 1]
```diff
@@ -12,7 +12,7 @@ readme = "README.md"
 python = ">=3.7"
 jsonschema = ">=3.2,<5.0"
 pandas = ">=1.0.0"
-duckdb = ">=0.6.0"
+duckdb = ">=0.8.0"
```
[the remark this answers is not in the document]
Comment (ADBond) [reply pyproject.toml:15]: If we're worried about user confusion if they end up with an older duckb and want to use DL we could consider adding a hook in the linker that checks duckdb version if settings contains any DL levels, and give a warning/instructions if so - similar to the ANSI-compliance thing in the Spark linker with date-casting

**Review by ADBond — APPROVED** [review]
Just a few small comments/discussion points but overall looks good to me 👍

File: splink/spark/spark_linker.py [hunk 2]
```diff
@@ -270,6 +270,11 @@ def _register_udfs_from_jar(self):
             ),
             ("jaccard", "uk.gov.moj.dash.linkage.JaccardSimilarity", DoubleType()),
             ("cosine_distance", "uk.gov.moj.dash.linkage.CosineDistance", DoubleType()),
+            (
+                "levdamerau_distance",
+                "uk.gov.moj.dash.linkage.LevDamerauDistance",
+                DoubleType(),
+            ),
```
[the remark this answers is not in the document]
Comment (RossKen) [reply splink/spark/spark_linker.py:277]: Yea, good point - will change it to match your naming 😉

**Pull request merged**